SampleApplications/SDK/Opc.Ua.Client.ComplexTypes/BaseComplexType.cs

```diff
@@ -0,0 +1,923 @@
+        {
+            return m_propertyList;
+        }
+        #endregion
+
+        #region Private Members
+        /// <summary>
+        /// Formatting helper.
+        /// </summary>
+        private void AddSeparator(StringBuilder body)
+        {
+            if (body.Length == 0)
+            {
+                body.Append("{");
+            }
+            else
+            {
+                body.Append(" | ");
+            }
+        }
+
+        /// <summary>
+        /// Append a property to the value string.
+        /// Handle arrays and enumerations.
+        /// </summary>
+        protected void AppendPropertyValue(
+            IFormatProvider formatProvider,
+            StringBuilder body,
+            object value,
+            int valueRank)
+        {
+            AddSeparator(body);
+            if (valueRank >= 0 &&
+                value is IEnumerable)
+            {
+                bool first = true;
+                var enumerable = value as IEnumerable;
+                body.Append("[");
+                foreach (var item in enumerable)
+                {
+                    if (!first)
+                    {
+                        body.Append(",");
+                    }
+                    AppendPropertyValue(formatProvider, body, item);
+                    first = false;
+                }
+                body.Append("]");
+            }
+            else
+            {
+                AppendPropertyValue(formatProvider, body, value);
+            }
+        }
+
+        /// <summary>
+        /// Append a property to the value string.
+        /// </summary>
+        private void AppendPropertyValue(
+            IFormatProvider formatProvider,
+            StringBuilder body,
+            object value)
+        {
+            if (value is byte[])
+            {
+                body.AppendFormat(formatProvider, "Byte[{0}]", ((byte[])value).Length);
+                return;
+            }
+
+            if (value is XmlElement)
+            {
+                body.AppendFormat(formatProvider, "<{0}>", ((XmlElement)value).Name);
+                return;
+            }
+
+            body.AppendFormat("{0}", value);
+        }
+
+        /// <summary>
+        /// Encode a property based on the property type and value rank.
+        /// </summary>
+        protected void EncodeProperty(IEncoder encoder, PropertyInfo property, int valueRank)
+        {
+            if (valueRank < 0)
+            {
+                EncodeProperty(encoder, property);
+            }
+            else
+            {
+                EncodePropertyArray(encoder, property);
+            }
+        }
+
+        /// <summary>
+        /// Encode a scalar property based on the property type.
+        /// </summary>
+        private void EncodeProperty(IEncoder encoder, PropertyInfo property)
+        {
+            var propertyType = property.PropertyType;
+            if (propertyType == typeof(Boolean))
```

Comment:
Minor: It would be great to have this as extension methods for IEncoder, i.e. Write<T>().      I have seen the same if/else in pub/sub.   The problem is that when we add more types (e.g. decimal), a lot of code needs to be touched.

Reply:
also for later ...